src/NHibernate/Transaction/AdoNetWithDistributedTransactionFactory.cs

```diff
@@ -94,90 +79,196 @@ public class DistributedTransactionContext : ITransactionContext, IEnlistmentNot
+				// phase would occur for this ressource.
+				var wasSuccessful = false;
+				try
 				{
-					logger.Debug("rolled back DTC transaction");
-					// Currently AfterTransactionCompletion is called by the handler for the TransactionCompleted event.
-					//sessionImplementor.AfterTransactionCompletion(false, null);
-					enlistment.Done();
-					IsInActiveTransaction = false;
+					wasSuccessful = e.Transaction.TransactionInformation.Status
+									== TransactionStatus.Committed;
 				}
+				catch (ObjectDisposedException ode)
+				{
+					_logger.Warn("Completed transaction was disposed, assuming transaction rollback", ode);
+				}
+				RunAfterTransactionActions(wasSuccessful);
 			}
 
-			void IEnlistmentNotification.InDoubt(Enlistment enlistment)
+			private volatile bool _afterTransactionActionDone;
+
+			private void RunAfterTransactionActions(bool wasSuccessful)
 			{
-				using (new SessionIdLoggingContext(sessionImplementor.SessionId))
+				if (_afterTransactionActionDone)
+					// Probably called from In-Doubt and TransactionCompleted.
+					return;
+				// Allow transaction completed actions to run while others stay blocked.
+				_bypassWait.Value = true;
+				try
 				{
-					sessionImplementor.AfterTransactionCompletion(false, null);
-					logger.Debug("DTC transaction is in doubt");
-					enlistment.Done();
-					IsInActiveTransaction = false;
+					using (new SessionIdLoggingContext(_sessionImplementor.SessionId))
+					{
+						// Flag active as false before running actions, otherwise the connection manager will refuse
+						// releasing the connection.
+						IsInActiveTransaction = false;
+						_sessionImplementor.AfterTransactionCompletion(wasSuccessful, null);
+						if (ShouldCloseSessionOnDistributedTransactionCompleted)
+						{
+							_sessionImplementor.CloseSessionFromDistributedTransaction();
+						}
+						_sessionImplementor.TransactionContext = null;
+					}
+				}
+				finally
+				{
+					_afterTransactionActionDone = true;
+					// Dispose releases blocked threads by the way.
+					// Must dispose in case !ShouldCloseSessionOnDistributedTransactionCompleted, since
+					// we nullify session TransactionContext, causing it to have nothing still holding it.
+					Dispose();
 				}
 			}
 
-			#endregion
+			private volatile bool _isDisposed;
 
 			public void Dispose()
 			{
-				if (AmbientTransation != null)
-					AmbientTransation.Dispose();
+				if (_isDisposed)
+					// Avoid disposing twice (happen when ShouldCloseSessionOnDistributedTransactionCompleted).
+					return;
+				_isDisposed = true;
+				Dispose(true);
+				GC.SuppressFinalize(this);
+			}
+
+			protected virtual void Dispose(bool disposing)
+			{
+				if (disposing)
+				{
+					if (AmbientTransation != null)
+					{
+						AmbientTransation.Dispose();
+						AmbientTransation = null;
+					}
+					_waitEvent.Set();
+					_waitEvent.Dispose();
```

Comment:
In my tests, disposing right after signaling does not cause an exception for thread potentially waiting. (I have run all test first without the catch around `WaitOne`, and do not got any failures, all tests passed.)